nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/eventhub/utils/ComponentStateCheckpointStore.java

```diff
@@ -0,0 +1,228 @@
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.BiFunction;
+
+public class ComponentStateCheckpointStore implements CheckpointStore {
+    public interface State {
+        StateMap getState() throws IOException;
+        boolean replaceState(StateMap oldValue, Map<String, String> newValue) throws IOException;
+    }
+
+    private final String identifier;
+    private final State state;
+
+    private static final String KEY_CHECKPOINT = "checkpoint";
+    private static final String KEY_OWNERSHIP = "ownership";
+
+    public ComponentStateCheckpointStore(String identifier, State state) {
+        this.identifier = identifier;
+        this.state = state;
+    }
+
+    public Flux<PartitionOwnership> listOwnership(String fullyQualifiedNamespace, String eventHubName, String consumerGroup) {
+        return Flux.defer(
+                () -> {
+                    Map<String, String> map;
+                    try {
+                        map = state.getState().toMap();
+                    } catch (IOException e) {
+                        return Flux.error(e);
+                    }
+                    List<PartitionOwnership> ownerships = getEntries(map, KEY_OWNERSHIP, this::convertOwnership);
+                    return Flux.fromIterable(ownerships).filter(
+                            ownership ->
+                                    ownership.getFullyQualifiedNamespace().equals(fullyQualifiedNamespace)
+                                    && ownership.getEventHubName().equals(eventHubName)
+                                    && ownership.getConsumerGroup().equals(consumerGroup)
+                    );
+                }
+        );
+    }
+
+    public Flux<PartitionOwnership> claimOwnership(List<PartitionOwnership> requestedPartitionOwnerships) {
+        return Flux.fromIterable(requestedPartitionOwnerships).flatMap(
+                partitionOwnership -> {
+                    final StateMap oldState;
+                    try {
+                        oldState = state.getState();
+                    } catch (IOException e) {
+                        return Mono.error(e);
+                    }
+
+                    String key = makeKey(
+                            KEY_OWNERSHIP,
+                            partitionOwnership.getFullyQualifiedNamespace(),
+                            partitionOwnership.getEventHubName(),
+                            partitionOwnership.getConsumerGroup(),
+                            partitionOwnership.getPartitionId()
+                    );
+
+                    final Map<String, String> newState = new HashMap<>(oldState.toMap());
+                    long timestamp = System.currentTimeMillis();
+                    String eTag = identifier + "/" + timestamp;
```

Comment:
Should the `eTag` be created based on the state version as opposed the current timestamp? That seems like it would align better with the purpose of the tag for tracking ownership and ensuring consistent version control across nodes.

Reply:
I'm not sure – but that logic was copied from the blob-based checkpoint store.

In fact, whenever I could, I have used the exact same logic between them.